### ISSUE TITLE
Bi-directional protocol commands

### ISSUE THREAD

**Issue by vcharpenay** [noflo/noflo-ui#443]
All Graph sub-protocol commands are sent from the UI to the runtime. Commands sent from the runtime to the UI are ignored. It prevents the runtime from performing automatic changes to the graph that is currently edited in the UI. Why not making these messages bi-directional? Use cases may comprise:

* automatically suggest links between existing nodes under certain conditions (`addedge` command)
* automatically add pre-configured nodes (e.g. with IIP) to a new graph (`addnode` command)
* reload a graph stored in the runtime (for instance following `component` messages sent after a `list` request).
* change edge color according to a runtime-specific logic, such as packet type going through the edge, not allowed wiring... (`changeedge` command)

And where are such graph commands likely to be processed within noflo-ui's code?

**Comment by jonnor** [comment]
They are intended to be bi-directional: The UI *should* reflect changes done on runtime side. It is just not implemented :)
There are two main components involved: [noflo-runtime](http://github.com/noflo/noflo-runtime)** for the protocol parts and [the-graph](http://github.com/the-grid/the-graph for the UI parts

* eventually fbp-runtime-client

**Comment by h0ru5** [comment]
I digged a bit into it and came up with this view:

it seems that 
https://github.com/noflo/noflo-runtime/blob/master/src/runtimes/base.coffee

does emit all incoming messages, so its just up to the handlers to do something with it.
So it would be somehing like a new component that does the reverse to his one:
https://github.com/noflo/noflo-runtime/blob/master/components/SendGraphChanges.coffee

I.e. it should listen to the runtime and apply the changes on the graph.
Then in nolfo-ui this component could be wired

sounds plausible?

**Comment by jonnor** [comment]
Yep, that is basically what is needed. Github.com/the-grid/The-graph
widget, which observes changes in graph instance light also need sone fixes

**Comment by forresto** [comment]
the-graph rerenders on `endTransaction`, which I assume would fire when getting changes from the runtime...?

**Comment by jonnor** [comment]
The connection code should be written liek that yes. So it could be that it will just-work!

**Comment by schmurfy** [comment]
is there any progress on this subject ?

**Comment by h0ru5** [comment]
There were efforts from @vcharpenay 
I think the master branch should be the HEAD state for this: https://github.com/vcharpenay/noflo-ui

**Comment by bergie** [comment]
@h0ru5 @vcharpenay interesting! would be nice to get a pull request on this

**Comment by chadrik** [comment]
@ifitzpatrick has done some work related to this. I'll let him comment.

**Comment by jonnor** [comment]
There was a PR proposed here, https://github.com/noflo/noflo-ui/pull/564
